File: frontend/src/app/validators/password.validator.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

export class PasswordValidator {
  // Password validation rules matching the backend PasswordValidator
  static strong(): ValidatorFn {
    return (control: AbstractControl): ValidationErrors | null => {
      const value = control.value;
      
      if (!value) {
        return null;
      }
      
      const errors: any = {};
      
      // Length validation (8-128 characters)
      if (value.length < 8) {
        errors.minLength = true;
      }
      
      if (value.length > 128) {
        errors.maxLength = true;
      }
      
      // Character type validation
      if (!/[A-Z]/.test(value)) {
        errors.uppercase = true;
      }
      
      if (!/[a-z]/.test(value)) {
        errors.lowercase = true;
      }
      
      if (!/[0-9]/.test(value)) {
        errors.digit = true;
      }
      
      if (!/[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]/.test(value)) {
        errors.specialChar = true;
      }
      
      // Check for common weak passwords
      const commonPasswords = [
        "password", "123456", "password123", "admin", "qwerty", 
        "letmein", "welcome", "monkey", "1234567890", "password1"
      ];
      
      const lowerValue = value.toLowerCase();
      if (commonPasswords.some(p => lowerValue === p)) {
        errors.commonPassword = true;
      }
      
      // Check for repeated characters (more than 3 consecutive)
      if (/(.)\1{3,}/.test(value)) {
        errors.repeatedChars = true;
      }
      
      // Check for sequential characters (like 123 or abc)
      let hasSequential = false;
      for (let i = 0; i <= value.length - 3; i++) {
        const substr = value.substring(i, i + 3);
        if (isSequential(substr)) {
          hasSequential = true;
          break;
        }
      }
      
      if (hasSequential) {
        errors.sequentialChars = true;
      }
      
      return Object.keys(errors).length > 0 ? errors : null;
    };
  }
  
  static match(passwordField: string, confirmPasswordField: string): ValidatorFn {
    return (control: AbstractControl): ValidationErrors | null => {
      const password = control.get(passwordField)?.value;
      const confirmPassword = control.get(confirmPasswordField)?.value;
      
      if (!password || !confirmPassword) {
        return null;
      }
      
      return password === confirmPassword ? null : { passwordMismatch: true };
    };
  }
}

function isSequential(str: string): boolean {
  // Check ascending sequence
  let isAscending = true;
  for (let i = 1; i < str.length; i++) {
    if (str.charCodeAt(i) !== str.charCodeAt(i - 1) + 1) {
      isAscending = false;
      break;
    }
  }
  
  // Check descending sequence
  let isDescending = true;
  for (let i = 1; i < str.length; i++) {
    if (str.charCodeAt(i) !== str.charCodeAt(i - 1) - 1) {
      isDescending = false;
      break;
    }
  }
  
  return isAscending || isDescending;
}

File: backend/src/main/java/com/dlvery/dlvery/util/PasswordValidator.java
package com.dlvery.dlvery.util;

import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

@Component
public class PasswordValidator {
    
    // Password validation patterns
    private static final Pattern UPPERCASE_PATTERN = Pattern.compile("[A-Z]");
    private static final Pattern LOWERCASE_PATTERN = Pattern.compile("[a-z]");
    private static final Pattern DIGIT_PATTERN = Pattern.compile("[0-9]");
    private static final Pattern SPECIAL_CHAR_PATTERN = Pattern.compile("[!@#$%^&*()_+\\-=\\[\\]{};':\"\\\\|,.<>\\/?]");
    
    // Password constraints
    private static final int MIN_PASSWORD_LENGTH = 8;
    private static final int MAX_PASSWORD_LENGTH = 128;
    
    // Common weak passwords to reject
    private static final String[] COMMON_WEAK_PASSWORDS = {
        "password", "123456", "password123", "admin", "qwerty", 
        "letmein", "welcome", "monkey", "1234567890", "password1"
    };
    
    public PasswordValidationResult validatePassword(String password) {
        List<String> errors = new ArrayList<>();
        
        if (password == null || password.isEmpty()) {
            errors.add("Password is required");
            return new PasswordValidationResult(false, errors);
        }
        
        // Length validation
        if (password.length() < MIN_PASSWORD_LENGTH) {
            errors.add("Password must be at least " + MIN_PASSWORD_LENGTH + " characters long");
        }
        
        if (password.length() > MAX_PASSWORD_LENGTH) {
            errors.add("Password must not exceed " + MAX_PASSWORD_LENGTH + " characters");
        }
        
        // Character type validation
        if (!UPPERCASE_PATTERN.matcher(password).find()) {
            errors.add("Password must contain at least one uppercase letter");
        }
        
        if (!LOWERCASE_PATTERN.matcher(password).find()) {
            errors.add("Password must contain at least one lowercase letter");
        }
        
        if (!DIGIT_PATTERN.matcher(password).find()) {
            errors.add("Password must contain at least one digit");
        }
        
        if (!SPECIAL_CHAR_PATTERN.matcher(password).find()) {
            errors.add("Password must contain at least one special character");
        }
        
        // Check for common weak passwords
        String lowerPassword = password.toLowerCase();
        for (String weakPassword : COMMON_WEAK_PASSWORDS) {
            if (lowerPassword.equals(weakPassword)) {
                errors.add("Password is too common and easily guessable");
                break;
            }
        }
        
        // Check for repeated characters (more than 3 consecutive)
        if (hasRepeatedCharacters(password, 3)) {
            errors.add("Password cannot contain more than 3 consecutive identical characters");
        }
        
        // Check for sequential characters (like 123 or abc)
        if (hasSequentialCharacters(password, 3)) {
            errors.add("Password cannot contain 3 or more sequential characters");
        }
        
        return new PasswordValidationResult(errors.isEmpty(), errors);
    }
    
    public PasswordValidationResult validatePasswordMatch(String password, String confirmPassword) {
        List<String> errors = new ArrayList<>();
        
        if (password == null || confirmPassword == null) {
            errors.add("Both password and confirmation are required");
            return new PasswordValidationResult(false, errors);
        }
        
        if (!password.equals(confirmPassword)) {
            errors.add("Passwords do not match");
        }
        
        return new PasswordValidationResult(errors.isEmpty(), errors);
    }
    
    private boolean hasRepeatedCharacters(String password, int maxRepeats) {
        for (int i = 0; i <= password.length() - maxRepeats; i++) {
            boolean isRepeated = true;
            char currentChar = password.charAt(i);

            for (int j = 1; j < maxRepeats; j++) {
                if (password.charAt(i + j) != currentChar) {
                    isRepeated = false;
                    break;
                }
            }

            if (isRepeated) return true;
        }

        return false;
    }
    
    private boolean hasSequentialCharacters(String password, int minSequence) {
        for (int i = 0; i <= password.length() - minSequence; i++) {
            boolean isAscending = true;
            boolean isDescending = true;

            for (int j = 1; j < minSequence; j++) {
                char current = password.charAt(i + j - 1);
                char next = password.charAt(i + j);

                if (next != current + 1) isAscending = false;
                if (next != current - 1) isDescending = false;
            }

            if (isAscending || isDescending) return true;
        }

        return false;
    }
    
    public static class PasswordValidationResult {
        private final boolean valid;
        private final List<String> errors;
        
        public PasswordValidationResult(boolean valid, List<String> errors) {
            this.valid = valid;
            this.errors = errors;
        }
        
        public boolean isValid() {
            return valid;
        }
        
        public List<String> getErrors() {
            return errors;
        }
        
        public String getMessage() {
            return String.join("; ", errors);
        }
    }
}